src/main/java/io/vertx/serviceproxy/generator/ServiceProxyGen.java

```diff
@@ -40,255 +54,277 @@ public String filename(ProxyModel model) {
+    return parseBlock("{" + methodBody + "}");
+  }
+
+  private String generateMethodBody(ProxyMethodInfo method) {
+    ParamInfo lastParam = utils.getLastParam(method);
     boolean hasResultHandler = utils.isResultHandler(lastParam);
+    StringBuilder methodBody = new StringBuilder();
+
     if (hasResultHandler) {
-      writer.code("if (closed) {\n");
-      writer.indent();
-      writer.stmt(lastParam.getName() + ".handle(Future.failedFuture(new IllegalStateException(\"Proxy is closed\")))");
-      if (method.isFluent())
-        writer.stmt("return this");
-      else
-        writer.stmt("return");
-      writer.unindent();
-      writer.code("}\n");
+      List<ParamInfo> paramsExcludedHandler = utils.paramsWithoutResultHandler(method);
+      methodBody.append(utils.generatedCodeBlock(
+        "if (closed) {\n"
+          + "  $L.handle(Future.failedFuture(new IllegalStateException(\"Proxy is closed\")));\n"
+          + "  $L;\n"
+          + "}\n",
+        lastParam.getName(), method.isFluent() ? "return this" : "return"));
+      methodBody.append(commonPart(method, paramsExcludedHandler));
+      methodBody.append(generateSendCallWithResultHandler(lastParam));
     } else {
-      writer.code("if (closed) throw new IllegalStateException(\"Proxy is closed\");\n");
+      List<ParamInfo> params = method.getParams().isEmpty() ? new ArrayList<>() : method.getParams();
+      methodBody.append(utils.generatedCodeBlock("if (closed) throw new IllegalStateException($S);", "Proxy is closed"));
+      methodBody.append(commonPart(method, params));
+      methodBody.append(utils.generatedCodeBlock("_vertx.eventBus().send(_address, _json, _deliveryOptions);"));
     }
-    if (method.isProxyClose())
-      writer.stmt("closed = true");
-    writer.stmt("JsonObject _json = new JsonObject()");
-    List<ParamInfo> paramsExcludedHandler =
-      (method.getParams().isEmpty()) ? new ArrayList<>() :
-        (hasResultHandler) ? method.getParams().subList(0, method.getParams().size() - 1) : method.getParams();
-    paramsExcludedHandler.forEach(p -> generateAddToJsonStmt(p, writer));
-    writer.newLine();
-    writer.stmt("DeliveryOptions _deliveryOptions = (_options != null) ? new DeliveryOptions(_options) : new DeliveryOptions()");
-    writer.stmt("_deliveryOptions.addHeader(\"action\", \"" + method.getName() + "\")");
-    if (hasResultHandler) {
-      generateSendCallWithResultHandler(lastParam, writer);
-    } else {
-      writer.stmt("_vertx.eventBus().send(_address, _json, _deliveryOptions)");
+
+    if (method.isProxyClose()) {
+      methodBody.append(utils.generatedCodeBlock("closed = true;"));
     }
+
+    return methodBody.toString();
   }
 
-  private void generateAddToJsonStmt(ParamInfo param, CodeWriter writer) {
+  private String commonPart(ProxyMethodInfo method, List<ParamInfo> paramsExcludedHandler) {
+    StringBuilder methodBody = new StringBuilder();
+    String addToJsonStatements = paramsExcludedHandler
```

Comment:
Collapse that in the usage place, avoid these variable declarations if you don't need them